src/main/java/com/fasterxml/jackson/databind/util/ClassUtil.java

```diff
@@ -1049,15 +1049,10 @@ public static boolean isEnumType(Class<?> rawType) {
      * class of enum instance (for "simple" enumerations), or its
      * superclass (for enums with instance fields or methods)
      */
-    @SuppressWarnings("unchecked")
     public static Class<? extends Enum<?>> findEnumType(Enum<?> en)
     {
         // enums with "body" are sub-classes of the formal type
-    	Class<?> ec = en.getClass();
-    	if (ec.getSuperclass() != Enum.class) {
-    	    ec = ec.getSuperclass();
-    	}
-    	return (Class<? extends Enum<?>>) ec;
+        return en.getDeclaringClass();
```

Comment:
I removed the loop, and added a case to `ClassUtilTest.testFindEnumType` which would have previously failed without it.